include/revng/Support/GraphAlgorithms.h

```diff
@@ -196,3 +201,215 @@ auto exitless_scc_range(NodeTy Entry) {
 
   return make_filter_range(Range, Filter);
 }
+
+namespace detail {
+
+template<typename T>
+T stripInverseHelper(llvm::Inverse<T> &&);
+
+template<typename T>
+T stripInverseHelper(T &&);
+
+template<typename T>
+using StripInverse = decltype(stripInverseHelper(std::declval<T>()));
+
+// clang-format off
+template<typename T>
+concept SupportedGraph = std::is_pointer_v<T> ||
+                         (StrictSpecializationOf<T, llvm::Inverse> &&
+                          std::is_pointer_v<StripInverse<T>>);
+// clang-format on
+
+template<typename GraphType>
+using GT = llvm::GraphTraits<GraphType>;
+
+} // namespace detail
+
+template<detail::SupportedGraph RawGraphType>
+class GraphNodeOrRootHelper {
+public:
+  using GraphType = detail::StripInverse<RawGraphType>;
+
+  using NodeRef = typename detail::GT<RawGraphType>::NodeRef;
+  using ValueType = std::variant<GraphType, NodeRef>;
+  using EquivalentPair = std::pair<GraphType, NodeRef>;
+
+  using GraphTrait = detail::GT<RawGraphType>;
+  using NodeTrait = std::conditional_t<std::is_same_v<GraphType, RawGraphType>,
+                                       detail::GT<NodeRef>,
+                                       detail::GT<llvm::Inverse<NodeRef>>>;
+
+  using Iterator = VariedIterator<typename GraphTrait::nodes_iterator,
+                                  typename NodeTrait::ChildIteratorType>;
+
+public:
+  ValueType Value;
+
+public:
+  GraphNodeOrRootHelper(GraphType Graph) : Value(Graph) {}
+  GraphNodeOrRootHelper(llvm::Inverse<GraphType> Inv) : Value(Inv.Graph) {}
+  GraphNodeOrRootHelper(NodeRef Node) : Value(std::move(Node)) {}
+  GraphNodeOrRootHelper(std::nullptr_t) : Value((GraphType) nullptr) {}
+  GraphNodeOrRootHelper(EquivalentPair P) : Value((GraphType) nullptr) {
+    if (P.first != nullptr)
+      Value = P.first;
+    else
+      Value = P.second;
+  }
```

Comment:
Why?

Reply:
Turns out, the `scc_iterator` uses this object inside `DenseMap`. Hence it needs the hashing trait and I don't feel comfortable supplying that one, so I got away with a way to convert this object to the closes object it provides traits to - a pair of pointers. Even if one of these two is always a `nullptr` (it's a waste of space), I feel like it's a usable trade-off for not having to worry about hashing of these objects.

Reply:
> Hence it needs the hashing trait

I suggest to 1) implement the hash trait, 2) build there a `pair` on the fly, 3) dispatch to the `pair`'s hash trait implementation.
This way we avoid implicit/explicit cast operators or extra constructors. They are suspicious without doc.

Reply:
Yep, this works :)

Reply:
Closing this, since I found a way to get rid of this without losing functionality.